File: contest/leetcode/linked-list-cycle.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

#include <cstdio>
struct ListNode {
    ListNode* next;
};

class Solution {
   public:
    bool hasCycle(ListNode* head) {
        ListNode *p0 = head, *p1 = head;
        while (p0 != NULL && p1 != NULL) {
            p0 = p0->next;
            p1 = p1->next;
            if (p1 == NULL) {
                return false;
            }
            p1 = p1->next;
            if (p0 == p1) {
                return true;
            }
        }
        return false;
    }
};

File: leetcode/linked-list-cycle.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head: return False
        x = head
        y = head.next
        while x and y:
            if x == y: return True
            x = x.next
            y = y.next
            if y: y = y.next
        return False
